File: src/edge-agent/app/src/main/kotlin/com/fccmiddleware/edge/adapter/common/IFccEventListener.kt
package com.fccmiddleware.edge.adapter.common

/**
 * Callback interface for unsolicited FCC events pushed over persistent connections.
 *
 * Used by DOMS TCP/JPL adapter to notify CadenceController of real-time events
 * without polling. Implementations must be thread-safe — callbacks may arrive
 * from the TCP read loop coroutine at any time.
 *
 * All callbacks are fire-and-forget from the adapter's perspective. Implementations
 * should not block or throw — errors must be caught and logged internally.
 */
interface IFccEventListener {

    /**
     * A pump's operational state has changed (e.g., IDLE -> CALLING -> DISPENSING).
     *
     * @param pumpNumber Canonical pump number (after offset adjustment).
     * @param newState The new canonical pump state.
     * @param fccStatusCode Raw vendor-specific status code for diagnostics.
     */
    fun onPumpStatusChanged(pumpNumber: Int, newState: PumpState, fccStatusCode: String?)

    /**
     * The FCC has signalled that one or more new transactions are available for retrieval.
     *
     * CadenceController should trigger an immediate fetchTransactions() call.
     *
     * @param notification Details about the available transaction(s).
     */
    fun onTransactionAvailable(notification: TransactionNotification)

    /**
     * A fuelling-in-progress update with live volume/amount data.
     *
     * Used for real-time display updates. Not stored — purely transient.
     *
     * @param pumpNumber Canonical pump number.
     * @param volumeMicrolitres Current dispensed volume in microlitres.
     * @param amountMinorUnits Current dispensed amount in minor currency units.
     */
    fun onFuellingUpdate(pumpNumber: Int, volumeMicrolitres: Long, amountMinorUnits: Long)

    /**
     * The persistent connection to the FCC has been lost unexpectedly.
     *
     * CadenceController should mark FCC as unreachable and trigger reconnect logic.
     *
     * @param reason Human-readable description of the disconnection cause.
     */
    fun onConnectionLost(reason: String)
}

/**
 * Notification that one or more transactions are available in the FCC buffer.
 *
 * @param fpId Fuelling point ID (vendor-specific pump identifier).
 * @param transactionBufferIndex Index in the FCC's supervised transaction buffer.
 * @param timestamp ISO 8601 UTC timestamp when the notification was received.
 */
data class TransactionNotification(
    val fpId: Int,
    val transactionBufferIndex: Int? = null,
    val timestamp: String,
)

File: src/desktop-edge-agent/src/FccDesktopAgent.Core/Adapter/Common/IFccEventListener.cs
namespace FccDesktopAgent.Core.Adapter.Common;

/// <summary>
/// Callback interface for unsolicited FCC events pushed over persistent connections.
///
/// Used by DOMS TCP/JPL adapter to notify the runtime controller of real-time events
/// without polling. Implementations must be thread-safe — callbacks may arrive
/// from the TCP read loop at any time.
/// </summary>
public interface IFccEventListener
{
    /// <summary>
    /// A pump's operational state has changed (e.g., IDLE -> CALLING -> DISPENSING).
    /// </summary>
    /// <param name="pumpNumber">Canonical pump number (after offset adjustment).</param>
    /// <param name="newState">The new canonical pump state.</param>
    /// <param name="fccStatusCode">Raw vendor-specific status code for diagnostics.</param>
    void OnPumpStatusChanged(int pumpNumber, PumpState newState, string? fccStatusCode);

    /// <summary>
    /// The FCC has signalled that one or more new transactions are available for retrieval.
    /// The controller should trigger an immediate FetchTransactionsAsync call.
    /// </summary>
    void OnTransactionAvailable(TransactionNotification notification);

    /// <summary>
    /// A fuelling-in-progress update with live volume/amount data (transient, not stored).
    /// </summary>
    void OnFuellingUpdate(int pumpNumber, long volumeMicrolitres, long amountMinorUnits);

    /// <summary>
    /// The persistent connection to the FCC has been lost unexpectedly.
    /// The controller should mark FCC as unreachable and trigger reconnect.
    /// </summary>
    void OnConnectionLost(string reason);
}

/// <summary>
/// Notification that one or more transactions are available in the FCC buffer.
/// </summary>
/// <param name="FpId">Fuelling point ID (vendor-specific pump identifier).</param>
/// <param name="TransactionBufferIndex">Index in the FCC's supervised transaction buffer.</param>
/// <param name="Timestamp">UTC timestamp when the notification was received.</param>
public sealed record TransactionNotification(
    int FpId,
    int? TransactionBufferIndex,
    DateTimeOffset Timestamp);
